packages/flutter/lib/src/foundation/logging.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:developer' as developer;
+
+/// A callback that, when evaluated, returns a log message.  Log messages must
+/// be encodable as JSON using `json.encode()`.
+typedef DebugLogMessageCallback = Object Function();
+
+/// Logs a message conditionally if the given identifying event [channel] is
+/// enabled (if `debugShouldLogEvent(channel)` is true).
+///
+/// Messages are obtained by evaluating [messageCallback] and must be encodable
+/// as JSON strings using `json.encode()`. In the event that logging is not
+/// enabled for the given [channel], [messageCallback] will not be evaluated.
+/// The cost of logging calls can be further mitigated at call sites by invoking
+/// them in a function that is only evaluated in debug mode. For example, to
+/// ignore logging in release mode you could wrap calls in an assert:
+///
+/// ```dart
+/// assert(() {
+///   debugLogEvent(logGestures, () => <String, int> {
+///    'x' : x,
+///    'y' : y,
+///    'z' : z,
+///   });
+/// }());
+///```
+///
+/// Logging for a given event channel can be enabled programmatically via
+/// [debugEnableLogging] or using a VM service call.
+///
+void debugLogEvent(LoggingChannel channel, DebugLogMessageCallback messageCallback) {
+  assert(channel != null);
+  if (!debugShouldLogEvent(channel)) {
+    return;
+  }
+
+  assert(messageCallback != null);
+  final Object message = messageCallback();
+  assert(message != null);
+
+  developer.log(json.encode(message), name: channel.name);
+}
+
+/// Identifies a logging channel.
+class LoggingChannel {
+  /// Singleton constructor. Calling `LoggingChannel(name)` returns the same
+  /// actual instance whenever it is called with the same string name.
+  factory LoggingChannel(String name, {String description}) =>
+      LoggingService.instance.getOrRegisterChannel(name, description ?? '');
```

Comment:
This API has a problem which is that the description that is used is the one that it was first called with. If you happen to look up the channel before you officially "create" the channel, you'll end up with the wrong description.

But also, the code flow here is weird. We have an official class `LoggingService` with a method `getOrRegisterChannel`, but we also have a `LoggingChannel` constructor. Why have both?

Also, in general this API and the `LoggingService` API have the problem that they don't do anything useful when you have a naming conflict. If two packages both try to register the same channel, but with very different semantics and different descriptions, they will collide and the user won't get any notification and then the IDE will produce garbage or crash or show errors because it can't interpret the data. This seems unfortunate.

Reply:
Some great points.

> If you happen to look up the channel before you officially "create" the channel, you'll end up with the wrong description.

In the envisioned world this feels unlikely to happen.  Especially if clients only register interest in channels that are returned from asking for the ones that exist.

> But also, the code flow here is weird. We have an official class LoggingService with a method getOrRegisterChannel, but we also have a LoggingChannel constructor. Why have both?

Ok, so I don't disagree but the book-keeping has to go somewhere and has to be accessible outside this library (so it can be queried from the service extension callback) and I wanted to make it as internal as possible.  I'd expect folks to not use the `LoggingService` at all and only interact w/ `LoggingChannels` and the 2 top level  methods (`debugLogEvent` and `shouldDebugLogEvent`).  Of course this is all motivated by the querying use case which I see some resistance to...

> Also, in general this API and the LoggingService API have the problem that they don't do anything useful when you have a naming conflict. If two packages both try to register the same channel, but with very different semantics and different descriptions, they will collide and the user won't get any notification and then the IDE will produce garbage or crash or show errors because it can't interpret the data. This seems unfortunate.

Right. We should handle collisions.  Just clobbering isn't good.  As for the likelihood of that happening, I suppose this is where naming conventions come in.  For package-contributed channels, I'd expect a `<package-name>.` prefix to suffice in practice.  (Needless to say, this should get documented well.)

Reply:
Even if we have the discovery mechanism, I think we can come up with a smaller API surface here. We don't need both the constructor and the registration mechanism to be public.

Similarly with the collisions, we should just assert that when you create it, it's not already there.

Reply:
> Even if we have the discovery mechanism, I think we can come up with a smaller API surface here. We don't need both the constructor and the registration mechanism to be public.

Great.  I have very little invested in the details to be perfectly honest.  If you could tell me what you'd like to see I'm happy to have a go at it.

> Similarly with the collisions, we should just assert that when you create it, it's not already there.

Agreed.  And document some naming conventions to ensure that it rarely happens...